File: feature/src/main/java/com/ortin/gesturetranslator/feature/managers/camera/CameraManagerImpl.java
package com.ortin.gesturetranslator.feature.managers.camera;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Matrix;
import android.media.Image;

import androidx.annotation.NonNull;
import androidx.camera.core.AspectRatio;
import androidx.camera.core.Camera;
import androidx.camera.core.CameraSelector;
import androidx.camera.core.ImageAnalysis;
import androidx.camera.core.ImageProxy;
import androidx.camera.lifecycle.ProcessCameraProvider;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.LifecycleOwner;

import com.google.common.util.concurrent.ListenableFuture;
import com.ortin.gesturetranslator.feature.managers.camera.listeners.CameraListener;
import com.ortin.gesturetranslator.feature.managers.camera.models.ImageFromCamera;
import com.ortin.gesturetranslator.feature.translators.YUVtoRGB;

public class CameraManagerImpl implements CameraManager {

    private final Context context;
    private ListenableFuture<ProcessCameraProvider> cameraProviderFuture;
    private final YUVtoRGB translator = new YUVtoRGB();
    private Bitmap bitmap;
    private Camera camera;

    public CameraManagerImpl(Context context) {
        this.context = context;
    }

    @Override
    public void loadImage(CameraListener cameraListener, LifecycleOwner lifecycleOwner) {
        cameraProviderFuture = ProcessCameraProvider.getInstance(context);
        cameraProviderFuture.addListener(new Runnable() {
            @Override
            public void run() {
                try {
                    ProcessCameraProvider cameraProvider = cameraProviderFuture.get();

                    ImageAnalysis imageAnalysis = new ImageAnalysis.Builder()
                            .setTargetAspectRatio(AspectRatio.RATIO_4_3)
                            .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
                            .build();

                    CameraSelector cameraSelector = new CameraSelector.Builder()
                            .requireLensFacing(CameraSelector.LENS_FACING_BACK)
                            .build();

                    imageAnalysis.setAnalyzer(ContextCompat.getMainExecutor(context), new ImageAnalysis.Analyzer() {
                        @Override
                        public void analyze(@NonNull ImageProxy image) {
                            Image img = image.getImage();
                            bitmap = translator.translateYUV(img, context);

                            Matrix matrix = new Matrix();
                            matrix.postRotate(image.getImageInfo().getRotationDegrees());
                            bitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.getWidth(), bitmap.getHeight(), matrix, true);

                            ImageFromCamera imageFromCamera = new ImageFromCamera(bitmap, image.getImageInfo().getRotationDegrees());
                            if (cameraListener != null) cameraListener.getImage(imageFromCamera);

                            image.close();
                        }
                    });

                    camera = cameraProvider.bindToLifecycle(lifecycleOwner, cameraSelector, imageAnalysis);

                } catch (Exception e) {
                    if (cameraListener != null) cameraListener.error(e);
                }
            }
        }, ContextCompat.getMainExecutor(context));
    }

    @Override
    public void setStatusFlashlight(boolean mode) {
        if (camera != null){
            camera.getCameraControl().enableTorch(mode);
        }
    }
}

File: feature/src/main/kotlin/com/ortin/gesturetranslator/feature/managers/camera/CameraManagerImpl.kt
package com.ortin.gesturetranslator.feature.managers.camera

import android.content.Context
import androidx.camera.core.Camera
import androidx.camera.core.CameraSelector
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.resolutionselector.AspectRatioStrategy
import androidx.camera.core.resolutionselector.ResolutionSelector
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.lifecycle.LifecycleOwner
import com.google.common.util.concurrent.ListenableFuture
import com.ortin.gesturetranslator.feature.managers.camera.analyzers.BitmapAnalyzer
import com.ortin.gesturetranslator.feature.managers.camera.listeners.CameraListener
import dagger.hilt.android.qualifiers.ApplicationContext
import java.util.concurrent.Executor
import java.util.concurrent.ExecutorService
import javax.inject.Inject

class CameraManagerImpl @Inject constructor(
    @ApplicationContext private val context: Context,
    private val cameraExecutor: ExecutorService,
    private val outputExecutor: Executor,
    private val cameraProviderFuture: ListenableFuture<ProcessCameraProvider>
) : CameraManager {
    private var camera: Camera? = null

    override fun loadImage(
        cameraListener: CameraListener,
        lifecycleOwner: LifecycleOwner,
        cameraFacing: Int
    ) {
        cameraProviderFuture.addListener({
            try {
                val cameraProvider = cameraProviderFuture.get()
                    ?: throw IllegalStateException("Camera initialization failed")

                val cameraSelector = CameraSelector.Builder()
                    .requireLensFacing(cameraFacing)
                    .build()

                val imageAnalysis = ImageAnalysis.Builder()
                    .setResolutionSelector(
                        ResolutionSelector.Builder()
                            .setAspectRatioStrategy(AspectRatioStrategy.RATIO_4_3_FALLBACK_AUTO_STRATEGY)
                            .build()
                    )
                    .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
                    .setOutputImageFormat(ImageAnalysis.OUTPUT_IMAGE_FORMAT_RGBA_8888)
                    .build()
                    .also {
                        it.setAnalyzer(cameraExecutor, BitmapAnalyzer(cameraFacing) { image ->
                            outputExecutor.execute {
                                cameraListener.getImage(image)
                            }
                        })
                    }

                cameraProvider.unbindAll()
                camera = cameraProvider.bindToLifecycle(
                    lifecycleOwner,
                    cameraSelector,
                    imageAnalysis
                )
            } catch (e: Exception) {
                cameraListener.error(e)
            }
        }, outputExecutor)
    }

    override fun setStatusFlashlight(mode: Boolean) {
        camera?.cameraControl?.enableTorch(mode)
    }
}
